### ISSUE TITLE
[BUG] Upgrade: rancher-monitoring charts can't be upgraded

### ISSUE THREAD

**Issue by bk201** [harvester/harvester#1983]
**Describe the bug**

This was observed after https://github.com/harvester/harvester-installer/pull/229.
`rancher-monitoring` and `rancher-monitoring-crd` ManagedCharts fail to upgrade.
The message is `another operation (install/upgrade/rollback) is in progress`.

```
fleet-local   mcc-rancher-monitoring                        0/1                       ErrApplied(1) [Cluster fleet-local/local: another operation (install/upgrade/rollback) is in progress]; mutatingwebhookconfiguration.admissionregistration.k8s.io rancher-monitoring-admission modified {"webhooks":[{"admissionReviewVersions":["v1","v1beta1"],"clientConfig":{"caBundle":"LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUJkVENDQVJ1Z0F3SUJBZ0lRUDd5MVA0K2cvRlZ5OGdaSW93bElDVEFLQmdncWhrak9QUVFEQWpBUE1RMHcKQ3dZRFZRUUtFd1J1YVd3eE1DQVhEVEl5TURFd05EQXlNRGcwTUZvWUR6SXhNakV4TWpFeE1ESXdPRFF3V2pBUApNUTB3Q3dZRFZRUUtFd1J1YVd3eE1Ga3dFd1lIS29aSXpqMENBUVlJS29aSXpqMERBUWNEUWdBRXEzQWhrdnh0CmVMUHBpOG45NmVkRmlWbGZiK0xRWHRmSVJFOUJDMmRlVmhxWjNvVUx1R0hucldqc3k0OWJMM0JVZEw0QWg4VEkKQUhORG9sbzNRVElGRHFOWE1GVXdEZ1lEVlIwUEFRSC9CQVFEQWdJRU1CTUdBMVVkSlFRTU1Bb0dDQ3NHQVFVRgpCd01CTUE4R0ExVWRFd0VCL3dRRk1BTUJBZjh3SFFZRFZSME9CQllFRlB3U0xOd2VaeE02VHZMS1dodnNxd2t5Cjc0RzNNQW9HQ0NxR1NNNDlCQU1DQTBnQU1FVUNJUUR3VmlHZEtSdXI3N2dUTElZVkxkNzZXM3N3eGtYd1I5c2gKdDBXN081azdNUUlnYmpRYWVsWEd6dnhQK1dtaHdGYWVndFdFVGJSSkQ2aXpzZVpJOWJ3c3ZQZz0KLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo=","service":{"name":"rancher-monitoring-operator","namespace":"cattle-monitoring-system","path":"/admission-prometheusrules/mutate","port":443}},"failurePolicy":"Ignore","matchPolicy":"Equivalent","name":"prometheusrulemutate.monitoring.coreos.com","namespaceSelector":{},"objectSelector":{},"reinvocationPolicy":"Never","rules":[{"apiGroups":["monitoring.coreos.com"],"apiVersions":["*"],"operations":["CREATE","UPDATE"],"resources":["prometheusrules"],"scope":"*"}],"sideEffects":"None","timeoutSeconds":10}]}; validatingwebhookconfiguration.admissionregistration.k8s.io rancher-monitoring-admission modified {"webhooks":[{"admissionReviewVersions":["v1","v1beta1"],"clientConfig":{"caBundle":"LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUJkVENDQVJ1Z0F3SUJBZ0lRUDd5MVA0K2cvRlZ5OGdaSW93bElDVEFLQmdncWhrak9QUVFEQWpBUE1RMHcKQ3dZRFZRUUtFd1J1YVd3eE1DQVhEVEl5TURFd05EQXlNRGcwTUZvWUR6SXhNakV4TWpFeE1ESXdPRFF3V2pBUApNUTB3Q3dZRFZRUUtFd1J1YVd3eE1Ga3dFd1lIS29aSXpqMENBUVlJS29aSXpqMERBUWNEUWdBRXEzQWhrdnh0CmVMUHBpOG45NmVkRmlWbGZiK0xRWHRmSVJFOUJDMmRlVmhxWjNvVUx1R0hucldqc3k0OWJMM0JVZEw0QWg4VEkKQUhORG9sbzNRVElGRHFOWE1GVXdEZ1lEVlIwUEFRSC9CQVFEQWdJRU1CTUdBMVVkSlFRTU1Bb0dDQ3NHQVFVRgpCd01CTUE4R0ExVWRFd0VCL3dRRk1BTUJBZjh3SFFZRFZSME9CQllFRlB3U0xOd2VaeE02VHZMS1dodnNxd2t5Cjc0RzNNQW9HQ0NxR1NNNDlCQU1DQTBnQU1FVUNJUUR3VmlHZEtSdXI3N2dUTElZVkxkNzZXM3N3eGtYd1I5c2gKdDBXN081azdNUUlnYmpRYWVsWEd6dnhQK1dtaHdGYWVndFdFVGJSSkQ2aXpzZVpJOWJ3c3ZQZz0KLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo=","service":{"name":"rancher-monitoring-operator","namespace":"cattle-monitoring-system","path":"/admission-prometheusrules/validate","port":443}},"failurePolicy":"Ignore","matchPolicy":"Equivalent","name":"prometheusrulemutate.monitoring.coreos.com","namespaceSelector":{},"objectSelector":{},"rules":[{"apiGroups":["monitoring.coreos.com"],"apiVersions":["*"],"operations":["CREATE","UPDATE"],"resources":["prometheusrules"],"scope":"*"}],"sideEffects":"None","timeoutSeconds":10}]}
fleet-local   mcc-rancher-monitoring-crd                    0/1                       ErrApplied(1) [Cluster fleet-local/local: another operation (install/upgrade/rollback) is in progress]
``` 

**To Reproduce**
1. Setup a 1.0.0 cluster.
2. Upgrade with master ISO.

**Expected behavior**

**Support bundle**

**Environment:**
 - Harvester ISO version: 
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630):

**Additional context**

**Comment by weihanglo** [comment]
Note that this might be related to https://github.com/rancher/fleet/issues/637.

**Comment by bk201** [comment]
I'm still seeing harvester chart goes into this state occasionally.
Even I already pause the managed chart and wait for all Rancehr stuff to settle down.
A trick to get rid of the state is to rollback the chart by using helm command. The managedchart will be applied again.

**Comment by bk201** [comment]
![image (2)](https://user-images.githubusercontent.com/1691518/159611116-39ef7940-e5c0-40b0-9e19-f40a1f048c5c.png)
@johnliu55tw also saw this after an upgrade.

**Comment by w13915984028** [comment]
I am debuging #2013, and encountered:
**In a single-node harvester with master-head ISO, without upgrade, this error is also observed.**

The main operation I am doing is:  
delete fleet-agent POD and set the new POD image to debug image.
`kubectl set image pod fleet-agent-8649b7d5b6-b5lzx -n cattle-fleet-local-system *=rancher/fleet-agent:dev`

**Suspect**: when previous fleet-agent has some helm operations ongoing, and the POD is deleted, then the new fleet-agent POD may complain `Error: another operation (install/upgrade/rollback) is in progress`

BundleDeployment
```
$ kubectl get bundledeployment -A 

NAMESPACE                                NAME                                          DEPLOYED                                                             MONITORED   STATUS
cluster-fleet-local-local-1a3d67d0a899   fleet-agent-local                             True                                                                 True        
cluster-fleet-local-local-1a3d67d0a899   local-managed-system-agent                    True                                                                 True        
cluster-fleet-local-local-1a3d67d0a899   mcc-harvester                                 True                                                                 True        configmap.v1 longhorn-system/longhorn-storageclass missing
cluster-fleet-local-local-1a3d67d0a899   mcc-harvester-crd                             True                                                                 True        
cluster-fleet-local-local-1a3d67d0a899   mcc-local-managed-system-upgrade-controller   True                                                                 True        
cluster-fleet-local-local-1a3d67d0a899   mcc-rancher-monitoring                        Error: another operation (install/upgrade/rollback) is in progress   True        mutatingwebhookconfiguration.admissionregistration.k8s.io rancher-monitoring-admission modified {"webhooks":[{"admissionReviewVersions":["v1","v1beta1"],"clientConfig":{"caBundle":"LS..tLQo=","service":{"name":"rancher-monitoring-operator","namespace":"cattle-monitoring-system","path":"/admission-prometheusrules/mutate","port":443}},"failurePolicy":"Ignore","matchPolicy":"Equivalent","name":"prometheusrulemutate.monitoring.coreos.com","namespaceSelector":{},"objectSelector":{},"reinvocationPolicy":"Never","rules":[{"apiGroups":["monitoring.coreos.com"],"apiVersions":["*"],"operations":["CREATE","UPDATE"],"resources":["prometheusrules"],"scope":"*"}],"sideEffects":"None","timeoutSeconds":10}]}
cluster-fleet-local-local-1a3d67d0a899   mcc-rancher-monitoring-crd                    True                                                                 True        

```
Bundle:  summarize status of related BundleDeployment.
![image](https://user-images.githubusercontent.com/31133476/159735544-685e98ca-f056-425e-b858-96790b5df118.png)

**Comment by w13915984028** [comment]
The realted debug info:

Helm relases are stored in kube secrets, the last release of rancher-monitoring, is in  `"status": "pending-upgrade"`.

```
kubectl get secrets -A | grep monit
cattle-monitoring-system                 sh.helm.release.v1.rancher-monitoring.v30                            helm.sh/release.v1                    1      3h49m

$ kubectl get secrets sh.helm.release.v1.rancher-monitoring.v30 -n cattle-monitoring-system -o JSON 
..
    },
    "kind": "Secret",
    "metadata": {
        "creationTimestamp": "2022-03-23T11:47:04Z",
        "labels": {
            "modifiedAt": "1648036025",
            "name": "rancher-monitoring",
            "owner": "helm",
            "status": "pending-upgrade",
            "version": "30"
        },
        "name": "sh.helm.release.v1.rancher-monitoring.v30",
        "namespace": "cattle-monitoring-system",
        "resourceVersion": "475385",
        "uid": "723f7644-694c-4371-aed7-24266da0a712"
    },
    "type": "helm.sh/release.v1"
}
```


```

	// errPending indicates that another instance of Helm is already applying an operation on a release.
	errPending = errors.New("another operation (install/upgrade/rollback) is in progress")

vendor/helm.sh/helm/v3/pkg/action/upgrade.go
	
// prepareUpgrade builds an upgraded release for an upgrade operation.
func (u *Upgrade) prepareUpgrade(name string, chart *chart.Chart, vals map[string]interface{}) (*release.Release, *release.Release, error) {
..

	// Concurrent `helm upgrade`s will either fail here with `errPending` or when creating the release with "already exists". This should act as a pessimistic lock.
	if lastRelease.Info.Status.IsPending() {
		return nil, nil, errPending
	}

```

**Comment by w13915984028** [comment]
A very possible reason:

When an fleet upgrade action (it calls helm source code API) executes more than 30 seconds, and the POD is terminated, the     on-going `"(install/upgrade/rollback)"` may be left there.

fleet-agent default termination time is 30s

$ kubectl get pods fleet-agent-8649b7d5b6-b5lzx -n cattle-fleet-local-system -o JSON
"terminationGracePeriodSeconds": 30,

helm master-head, has difference with the embedded helm used in fleet master-head

**Helm will release the  StatusPendingUpgrade release when termination.**  

![image](https://user-images.githubusercontent.com/31133476/159791581-dfe7c493-5710-4295-a3d6-fac04dc2b606.png)


![image](https://user-images.githubusercontent.com/31133476/159791770-3108ad4b-c33d-4c5a-97dd-e0ac9ff6548b.png)


@yasker @guangbochen @rebeccazzzz

**Comment by w13915984028** [comment]
The overall process of /rancher/fleet/./vendor/helm.sh/helm/v3/pkg/action/upgrade.go is:

```

func (u *Upgrade) Run
	u.cfg.Log("preparing upgrade for %s", name)
	currentRelease, upgradedRelease, err := u.prepareUpgrade(name, chart, vals)  ---------- prepare  release.Release, set status as release.StatusPendingUpgrade

	u.cfg.Log("performing update for %s", name)
	res, err := u.performUpgrade(currentRelease, upgradedRelease)  ------------------------ perform upgrade

		u.cfg.Log("creating upgraded release for %s", upgradedRelease.Name)
	  if err := u.cfg.Releases.Create(upgradedRelease); err != nil { ------------------------- create new release

```


**BUG:  when POD is terminated here after running over 30 seconds, or the POD is force terminated, the created upgradedRelease is left over, it blocks following upgrade operations, unless user deletes the related release manually, or user helm command to recover** 
```

		originalRelease.Info.Status = release.StatusSuperseded --------------------------------- previous version is set to superseded
		u.cfg.recordRelease(originalRelease)

		upgradedRelease.Info.Status = release.StatusDeployed  ---------------------------------- update to deployed
		if len(u.Description) > 0 {

	if !u.DryRun {
		u.cfg.Log("updating status for upgraded release for %s", name)
		if err := u.cfg.Releases.Update(upgradedRelease); err != nil {

```

log taken from fleet-agent:
```
time="2022-03-16T20:23:14Z" level=info msg="performing update for fleet-agent-local"
time="2022-03-16T20:23:14Z" level=info msg="creating upgraded release for fleet-agent-local"
time="2022-03-16T20:23:14Z" level=info msg="updating status for upgraded release for harvester"
time="2022-03-16T20:23:14Z" level=info msg="getting history for release harvester"
time="2022-03-16T20:23:14Z" level=info msg="preparing upgrade for harvester-crd"
time="2022-03-16T20:23:14Z" level=info msg="updating status for upgraded release for fleet-agent-local"
```


The `left` release:
```
$ kubectl get secrets sh.helm.release.v1.rancher-monitoring.v30 -n cattle-monitoring-system -o JSON 
..
    },
    "kind": "Secret",
    "metadata": {
        "creationTimestamp": "2022-03-23T11:47:04Z",
        "labels": {
            "modifiedAt": "1648036025",
            "name": "rancher-monitoring",
            "owner": "helm",
            "status": "pending-upgrade",
```

**Comment by w13915984028** [comment]
When checking the fleet-agent log, it showes, even in a normal running Harvester cluster there are also many "performing update for ... " log.   (Why it does so in under investigation via #2013 .)

There are chances to hit this bug.

**Comment by bk201** [comment]
@w13915984028 @weihanglo Do you know any workaround to get rid of this state? I did have luck to rollback the revision on a chart before, not sure if there is a better way.

**Comment by w13915984028** [comment]
@bk201 I found one possible cause, described in https://github.com/harvester/harvester/issues/1983#issuecomment-1076812969 , but not sure if fleet-agent noticed this and has plan to improve it.

**Comment by w13915984028** [comment]
The related fleet-agent issue: https://github.com/rancher/fleet/issues/637 , but looks no related fix.

**Comment by bk201** [comment]
Use also [hit this issue](https://github.com/harvester/harvester/issues/2280#issuecomment-1153203729) and can't apply workaround: 

> Getting same error upgrading from 1.0 to 1.0.1. Can't role back helm chart.
> 
> `helm history rancher-monitoring -n cattle-monitoring-system
> 
> REVISION UPDATED STATUS CHART APP VERSION DESCRIPTION
> 1 Tue Mar 15 20:18:48 2022 superseded rancher-monitoring-100.0.0+up16.6.0 0.48.0 superseded by new release`

**Comment by felixtegelkamp** [comment]
My workaround to complete upgrade.

1. Check bundles
   ` kubectl get bundles -A`
    <img width="1333" alt="Bildschirmfoto 2022-09-27 um 23 53 35" src="https://user-images.githubusercontent.com/9285028/192646323-0b4ee91a-5864-4da2-8920-02a49e87f8a4.png">
2. And also check secrets as helm releases are stored in here
    `kubectl get secrets -A -l  status=pending-upgrade`
    <img width="830" alt="Bildschirmfoto 2022-09-27 um 23 53 58" src="https://user-images.githubusercontent.com/9285028/192646630-e2c95253-3f89-4967-b0bd-7e18514af3a4.png">
3. Seems helm deployment rancher-monitoring and the harvester deployment are stuck somehow.
4. Verified helm status and rolled back rancher monitoring
    ``` sh
     helm status rancher-monitoring -n cattle-monitoring-system 
     helm history rancher-monitoring -n cattle-monitoring-system
     helm rollback rancher-monitoring -n cattle-monitoring-system
    ```
    <img width="1194" alt="Bildschirmfoto 2022-09-27 um 23 55 35" src="https://user-images.githubusercontent.com/9285028/192646973-292cda99-f7ad-46e8-8946-abf262a18352.png">
    <img width="1194" alt="Bildschirmfoto 2022-09-27 um 23 56 02" src="https://user-images.githubusercontent.com/9285028/192647091-b8e96d30-3c64-4fda-b0cc-8ab91aac236e.png">
5. Verify and Rollback harvester
    ```sh
    helm status harvester -n harvester-system
    helm history harvester -n harvester-system
    helm rollback harvester -n harvester-system
    ```
    <img width="894" alt="Bildschirmfoto 2022-09-27 um 23 57 25" src="https://user-images.githubusercontent.com/9285028/192647619-7597d612-4f11-4e62-81f6-64e9d313d6c2.png">

The upgrade process will continue shortly after harvester rollback is completed.

**Comment by w13915984028** [comment]
update @ 2022.10.06:

`Harvester v1.1.0-rc2`  has `fleet v0.4.0-rc3`, which in turn has `helm v3.9.0`, that includes the `context cancel`, and should have fixed the issue.

@bk201 @guangbochen We can move this to `read-for-test`, and keep an eye if this issue is reported again.



```
harv41:~ # kk get pods -n cattle-fleet-local-system fleet-agent-5b46776c54-v76f6 -o YAML | grep image
    image: rancher/fleet-agent:v0.4.0-rc3
```

![image](https://user-images.githubusercontent.com/31133476/194416255-d535d924-0447-480f-8ee8-44bd840c3909.png)

https://github.com/rancher/fleet/commit/eae202f68985ce20ff8f97586cd31ee64e7df020
![image](https://user-images.githubusercontent.com/31133476/194416349-cee25f79-51ba-44fd-9480-f2ae39ad854e.png)

https://github.com/helm/helm/blob/release-3.9/pkg/action/upgrade.go#L263
```func (u *Upgrade) performUpgrade(ctx context.Context, originalRelease, upgradedRelease *release.Release) (*release.Release, error) {```
![image](https://user-images.githubusercontent.com/31133476/194416883-b6ae06cb-893d-456c-b807-a87bd057bece.png)

**Comment by bk201** [comment]
Thanks, @w13915984028 , we can keep an eye on it.
Upgrading Rancher and fleet is the first stop in the upgrade process, so the fix should apply to all managed charts after it.

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.1.0-rc3`. Close this issue.

### Result
The failed to upgrade rancher-monitoring chart did not happen on the latest upgrade from v1.0.3 to v1.1.0-rc3

* Can successfully upgrade from v1.0.3 to v1.1.0-rc3
  ![image](https://user-images.githubusercontent.com/29251855/196444374-c04e7097-b968-4249-b841-a640890fb222.png)

* rancher-monitoring and rancher-monitoring-crd ManagedCharts have no failure message after upgarding to v1.1.0-rc3
  ```
  n1-103upd:~ # kubectl get bundledeployment -A
  NAMESPACE                                NAME                                          DEPLOYED   MONITORED   STATUS
  cluster-fleet-local-local-1a3d67d0a899   fleet-agent-local                             True       True        
  cluster-fleet-local-local-1a3d67d0a899   local-managed-system-agent                    True       True        
  cluster-fleet-local-local-1a3d67d0a899   mcc-harvester                                 True       True        
  cluster-fleet-local-local-1a3d67d0a899   mcc-harvester-crd                             True       True        
  cluster-fleet-local-local-1a3d67d0a899   mcc-local-managed-system-upgrade-controller   True       True        
  cluster-fleet-local-local-1a3d67d0a899   mcc-rancher-logging                           True       True        
  cluster-fleet-local-local-1a3d67d0a899   mcc-rancher-logging-crd                       True       True        
  cluster-fleet-local-local-1a3d67d0a899   mcc-rancher-monitoring                        True       True        
  cluster-fleet-local-local-1a3d67d0a899   mcc-rancher-monitoring-crd                    True       True
  ```
* rancher monitoring chart upgraded to `fleet-agent:v0.4.0`
  ```
  n1-103upd:~ # kubectl get pods -n cattle-fleet-local-system
  NAME                           READY   STATUS    RESTARTS   AGE
  fleet-agent-58fd796595-jpt8v   1/1     Running   0          90m
  
  n1-103upd:~ # kubectl get pods -n cattle-fleet-local-system fleet-agent-58fd796595-jpt8v -o YAML | grep image
      image: rancher/fleet-agent:v0.4.0
      imagePullPolicy: IfNotPresent
      image: docker.io/rancher/fleet-agent:v0.4.0
      imageID: sha256:fa99327e2bd8f4ed736c5e5c8bab1678b34cced1e8ff308414302627600d18fb
  ```

### Test Information
* Test Environment: 4 nodes harvester bare machines
* Harvester version: `v1.1.0-rc3`

### Verify Steps
1. Prepare a 4 nodes v1.0.3 Harvester cluster 
2. Install several images
3. Create three VMs
4. Enable Network 
5. Create vlan1 network
6. Shutdown all VMs
7. Upgrade to v1.1.0-rc3
